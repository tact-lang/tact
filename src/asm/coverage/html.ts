import {
    CoverageSummary,
    generateCoverageSummary,
    Line,
} from "@/asm/coverage/data";
import {
    MAIN_TEMPLATE,
    SUMMARY_TEMPLATE,
} from "@/asm/coverage/templates/templates";

const templates = {
    main: MAIN_TEMPLATE,
    summary: SUMMARY_TEMPLATE,
};

const renderTemplate = (
    template: string,
    data: Record<string, unknown>,
): string => {
    return template.replaceAll(/{{(\w+)}}/g, (_, key) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        return data[key]?.toString() ?? "";
    });
};

const formatGasCosts = (gasCosts: readonly number[]): string => {
    if (gasCosts.length === 0) return "";
    if (gasCosts.length === 1) return gasCosts[0]?.toString() ?? "0";

    const gasCount: Map<number, number> = new Map();
    for (const gas of gasCosts) {
        gasCount.set(gas, (gasCount.get(gas) ?? 0) + 1);
    }

    if (gasCount.size === 1) {
        const firstEntry = [...gasCount.entries()][0];
        return firstEntry?.[0]?.toString() ?? "";
    }

    return [...gasCount.entries()]
        .sort(([gas1], [gas2]) => gas1 - gas2)
        .map(([gas, count]) => `${gas} x${count}`)
        .join(", ");
};

export const calculateTotalGas = (gasCosts: readonly number[]): number => {
    return gasCosts.reduce((sum, gas) => sum + gas, 0);
};

const generateLineHtml = (
    line: Line,
    index: number,
    maxGasPerLine: number,
    totalGas: number,
): string => {
    const lineNumber = index + 1;
    const className = line.info.$;

    let gasHtml = `<div class="gas"></div>`;
    let hitsHtml = `<div class="hits"></div>`;
    let gasPercentStyle = "";

    if (line.info.$ === "Covered") {
        const gasInfo = line.info.gasCosts;
        const detailedGasCost = formatGasCosts(gasInfo);
        const totalGasCost = calculateTotalGas(gasInfo);

        const gasPercentage = Math.sqrt(totalGasCost / maxGasPerLine) * 100;
        const totalGasPercentage = (totalGasCost / totalGas) * 100;

        gasPercentStyle = ` style="--gas-percent:${gasPercentage.toFixed(4)}%" data-gas-percent="${totalGasPercentage.toFixed(2)}%"`;

        gasHtml = `<div class="gas">
            <span class="gas-detailed">${detailedGasCost}</span>
            <span class="gas-sum">${totalGasCost}</span>
        </div>`;
        hitsHtml = `<div class="hits" title="Number of times executed">${line.info.hits}</div>`;
    }

    return `<div class="line ${className}" id="L${lineNumber}"${gasPercentStyle} data-line-number="${lineNumber}">
    <div class="line-number">${lineNumber}</div>
    ${gasHtml}
    ${hitsHtml}
    <pre>${line.line.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;").replaceAll("'", "&#039;")}</pre>
</div>`;
};

const generateInstructionRowsHtml = (summary: CoverageSummary): string => {
    return summary.instructionStats
        .map((stat) => {
            const percentValue = (stat.totalGas / summary.totalGas) * 100;
            return `<tr>
                <td data-value="${stat.name}"><code>${stat.name}</code></td>
                <td data-value="${stat.totalGas}">${stat.totalGas}</td>
                <td data-value="${stat.totalHits}">${stat.totalHits}</td>
                <td data-value="${stat.avgGas}">${stat.avgGas}</td>
                <td data-value="${percentValue}">
                    <div class="percent-container">
                        <div class="percent-text">${percentValue.toFixed(2)}%</div>
                        <div class="percent-bar">
                            <div class="percent-fill" style="width: ${percentValue}%"></div>
                        </div>
                    </div>
                </td>
            </tr>`;
        })
        .join("\n");
};

export const generateHtml = (lines: readonly Line[]): string => {
    const summary = generateCoverageSummary(lines);

    const maxGas = Math.max(
        ...lines.map((line) =>
            line.info.$ === "Covered"
                ? line.info.gasCosts.reduce((sum, gas) => sum + gas, 0)
                : 0,
        ),
    );

    const htmlLines = lines
        .map((line, index) =>
            generateLineHtml(line, index, maxGas, summary.totalGas),
        )
        .join("\n");

    const templateData = {
        coverage_percentage: summary.coveragePercentage.toFixed(2),
        covered_lines: summary.coveredLines,
        total_lines: summary.totalLines,
        total_gas: summary.totalGas,
        total_hits: summary.totalHits,
        instruction_rows: generateInstructionRowsHtml(summary),
    };

    const summaryHtml = renderTemplate(templates.summary, templateData);

    return renderTemplate(templates.main, {
        SUMMARY_CONTENT: summaryHtml,
        CODE_CONTENT: htmlLines,
    });
};
