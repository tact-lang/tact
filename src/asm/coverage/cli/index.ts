import { Cell } from "@ton/core";
import { readFileSync, writeFileSync } from "node:fs";
import { generateHtml } from "@/asm/coverage/html";
import * as path from "node:path";
import * as fs from "node:fs";
import { collectAsmCoverage, collectFuncCoverage } from "@/asm/coverage/index";
import { CoverageSummary } from "@/asm/coverage/data";

const USAGE =
    "Usage: coverage <boc-file-path> <log-file-path> [<func-source-path> <func-mapping-path>]";

const main = () => {
    const args = process.argv.slice(2);

    const bocFilePath = args.at(0);
    const logPath = args.at(1);
    const funcSources = args.at(2);
    const funcMappingPath = args.at(3);
    const outputDir = "./";

    if (bocFilePath === undefined || logPath === undefined) {
        console.log(USAGE);
        process.exit(1);
    }

    console.log("âŒ› Generating coverage reports...");

    const cell = Cell.fromBoc(readFileSync(bocFilePath))[0];
    if (!cell) {
        console.error("Cannot parse BoC");
        process.exit(1);
    }

    const logs = readFileSync(logPath, "utf8");

    if (funcSources !== undefined && funcMappingPath !== undefined) {
        const { lines, summary } = collectFuncCoverage(
            cell,
            logs,
            funcSources,
            funcMappingPath,
        );

        printSummary("func", summary);

        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }

        const combinedReportPath = path.join(outputDir, "coverage-func.html");
        writeFileSync(combinedReportPath, generateHtml(lines));
        console.log(`\nâœ… Report generated: ${combinedReportPath}`);
    }

    {
        const { lines, summary } = collectAsmCoverage(cell, logs);

        printSummary("asm", summary);

        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }

        const combinedReportPath = path.join(outputDir, "coverage-asm.html");
        writeFileSync(combinedReportPath, generateHtml(lines));
        console.log(`\nâœ… Report generated: ${combinedReportPath}`);
    }
};

const printSummary = (kind: "func" | "asm", summary: CoverageSummary) => {
    const forKind = kind === "func" ? "FUNC CODE" : "ASSEMBLY CODE";
    const linesKind = kind === "func" ? "Lines" : "Instructions";

    console.log(`\nðŸ“Š COVERAGE SUMMARY FOR ${forKind}`);
    console.log("==================");
    console.log(`Coverage: ${summary.coveragePercentage.toFixed(2)}%`);
    console.log(`${linesKind}: ${summary.coveredLines}/${summary.totalLines}`);
    console.log(`Total Gas Used: ${summary.totalGas}`);
    console.log(`Executed ${linesKind}: ${summary.totalHits}`);
};

main();
