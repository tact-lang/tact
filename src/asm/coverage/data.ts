import { Step, TraceInfo } from "@/asm/trace";

export type Line = {
    readonly line: string;
    readonly info: Covered | Uncovered | Skipped;
};

export type Covered = {
    readonly $: "Covered";
    readonly hits: number;
    readonly gasCosts: readonly number[];
};

export type Uncovered = {
    readonly $: "Uncovered";
};

export type Skipped = {
    readonly $: "Skipped";
};

export type InstructionStat = {
    readonly name: string;
    readonly totalGas: number;
    readonly totalHits: number;
    readonly avgGas: number;
};

export type CoverageSummary = {
    readonly totalLines: number;
    readonly coveredLines: number;
    readonly uncoveredLines: number;
    readonly coveragePercentage: number;
    readonly totalGas: number;
    readonly totalHits: number;
    readonly instructionStats: readonly InstructionStat[];
};

export const buildLineInfo = (
    trace: TraceInfo,
    asm: string,
): readonly Line[] => {
    const lines = asm.split("\n");

    const perLineSteps: Map<number, Step[]> = new Map();

    for (const step of trace.steps) {
        if (step.loc === undefined) continue;
        const line = step.loc.line;

        perLineSteps.set(line + 1, [
            ...(perLineSteps.get(line + 1) ?? []),
            step,
        ]);

        if (step.loc.otherLines.length > 0) {
            for (const otherLine of step.loc.otherLines) {
                perLineSteps.set(otherLine + 1, [
                    ...(perLineSteps.get(otherLine + 1) ?? []),
                    step,
                ]);
            }
        }
    }

    return lines.map((line, idx): Line => {
        const info = perLineSteps.get(idx + 1);
        if (info) {
            const gasInfo = info.map((step) => normalizeGas(step.gasCost));

            return {
                line,
                info: {
                    $: "Covered",
                    hits: gasInfo.length,
                    gasCosts: gasInfo,
                },
            };
        }

        if (!isExecutableLine(line)) {
            return {
                line,
                info: {
                    $: "Skipped",
                },
            };
        }

        return {
            line,
            info: {
                $: "Uncovered",
            },
        };
    });
};

export const buildFuncLineInfo = (
    traces: TraceInfo[],
    funcCode: string,
): Line[] => {
    const lines = funcCode.split("\n");

    const perLineStepsArray: Map<number, Step[][]> = new Map();

    for (const trace of traces) {
        const perLineSteps: Map<number, Step[]> = new Map();
        for (const step of trace.steps) {
            if (step.funcLoc === undefined) continue;
            const line = step.funcLoc.line;

            perLineSteps.set(line, [...(perLineSteps.get(line) ?? []), step]);
        }

        for (const [line, steps] of perLineSteps.entries()) {
            const perLineStep = perLineStepsArray.get(line);
            if (perLineStep === undefined) {
                perLineStepsArray.set(line, [steps]);
            } else {
                perLineStep.push(steps);
            }
        }
    }

    return lines.map((line, idx): Line => {
        const infos = perLineStepsArray.get(idx + 1);
        if (infos) {
            const gasInfo = infos.flatMap((it) =>
                it.map((step) => normalizeGas(step.gasCost)),
            );

            return {
                line,
                info: {
                    $: "Covered",
                    hits: infos.length,
                    gasCosts: gasInfo,
                },
            };
        }

        if (!isExecutableLine(line)) {
            return {
                line,
                info: {
                    $: "Skipped",
                },
            };
        }

        return {
            line,
            info: {
                $: "Uncovered",
            },
        };
    });
};

const normalizeGas = (gas: number): number => {
    if (gas > 10000) {
        return 26;
    }
    return gas;
};

export const isExecutableLine = (line: string): boolean => {
    const trimmed = line.trim();
    return (
        !trimmed.includes("=>") && // dictionary
        trimmed !== "}" && // close braces
        !trimmed.includes("} {") && // IFREFELSEREF
        !trimmed.includes(";;") && // FunC comment line
        !trimmed.includes("#pragma ") &&
        !trimmed.includes("#include") &&
        !trimmed.includes("inline {") && // function signature
        !trimmed.includes("inline;") && // function signature
        !trimmed.includes("inline_ref {") && // function signature
        !trimmed.includes("inline_ref;") && // function signature
        !trimmed.includes(" method_id {") && // method signature
        !trimmed.includes('asm """') && // function signature
        !trimmed.includes('asm "') && // function signature
        !trimmed.includes("asm(") && // function signature
        trimmed.length > 0
    );
};

export const generateCoverageSummary = (
    lines: readonly Line[],
): CoverageSummary => {
    const totalExecutableLines = lines.filter((line) =>
        isExecutableLine(line.line),
    ).length;

    const coveredLines = lines.filter(
        (line) => isExecutableLine(line.line) && line.info.$ === "Covered",
    ).length;
    const uncoveredLines = totalExecutableLines - coveredLines;
    const coveragePercentage = (coveredLines / totalExecutableLines) * 100;

    let totalGas = 0;
    let totalHits = 0;

    const instructionMap: Map<
        string,
        { readonly totalGas: number; readonly hits: number }
    > = new Map();

    for (const line of lines) {
        if (line.info.$ !== "Covered") continue;

        const lineGas = line.info.gasCosts.reduce((sum, gas) => sum + gas, 0);
        totalGas += lineGas;
        totalHits += line.info.hits;
        const trimmedLine = line.line.trim();
        const instructionName = trimmedLine.split(/\s+/)[0];
        if (instructionName !== undefined) {
            const current = instructionMap.get(instructionName) ?? {
                totalGas: 0,
                hits: 0,
            };
            instructionMap.set(instructionName, {
                totalGas: current.totalGas + lineGas,
                hits: current.hits + line.info.hits,
            });
        }
    }

    const instructionStats: InstructionStat[] = [...instructionMap.entries()]
        .map(([name, stats]) => ({
            name,
            totalGas: stats.totalGas,
            totalHits: stats.hits,
            avgGas: Math.round((stats.totalGas / stats.hits) * 100) / 100,
        }))
        .sort((a, b) => b.totalGas - a.totalGas);

    return {
        totalLines: totalExecutableLines,
        coveredLines,
        uncoveredLines,
        coveragePercentage,
        totalGas,
        totalHits,
        instructionStats,
    };
};
