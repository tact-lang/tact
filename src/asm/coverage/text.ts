import { CoverageSummary, InstructionStat, Line } from "@/asm/coverage/data";
import { calculateTotalGas } from "@/asm/coverage/html";

export const generateShortSummary = (summary: CoverageSummary): string => {
    return [
        "Coverage Summary:",
        `Lines: ${summary.coveredLines}/${summary.totalLines} (${summary.coveragePercentage.toFixed(2)}%)`,
        `Total Gas: ${summary.totalGas}`,
        `Total Hits: ${summary.totalHits}`,
    ].join("\n");
};

export const generateTextReport = (
    lines: readonly Line[],
    summary: CoverageSummary,
): string => {
    const maxLineNumberWidth = lines.length.toString().length;

    const annotatedLines = lines
        .map((line, index) => {
            const { gasInfo, hitsInfo, status } = lineInfo(line);

            const lineNumber = index + 1;
            const lineNumberPres = lineNumber
                .toString()
                .padStart(maxLineNumberWidth);

            return `${lineNumberPres} ${status}| ${line.line.padEnd(40)} |${gasInfo.padEnd(10)} |${hitsInfo}`;
        })
        .join("\n");

    const summaryText = [
        "Coverage Summary:",
        `Lines: ${summary.coveredLines}/${summary.totalLines} (${summary.coveragePercentage.toFixed(2)}%)`,
        `Total Gas: ${summary.totalGas}`,
        `Total Hits: ${summary.totalHits}`,
        "",
        "Instruction Stats:",
        ...instructionsStats(summary),
    ].join("\n");

    return `${summaryText}\n\nAnnotated Code:\n${annotatedLines}`;
};

type LineInfo = {
    readonly gasInfo: string;
    readonly hitsInfo: string;
    readonly status: string;
};

const lineInfo = (line: Line): LineInfo => {
    if (line.info.$ === "Covered") {
        const totalGas = calculateTotalGas(line.info.gasCosts);
        const gasInfo = ` gas:${totalGas}`;
        const hitInfo = ` hits:${line.info.hits}`;
        return { gasInfo, hitsInfo: hitInfo, status: "✓ " };
    }

    if (line.info.$ === "Uncovered") {
        return { gasInfo: "", hitsInfo: "", status: "✗ " };
    }

    return { gasInfo: "", hitsInfo: "", status: "  " };
};

const instructionsStats = (summary: CoverageSummary) =>
    summary.instructionStats.map((stat) =>
        formatInstructionStat(stat, summary.totalGas),
    );

const formatInstructionStat = (stat: InstructionStat, totalGas: number) => {
    const name = stat.name.padEnd(15);
    const totalGasStr = stat.totalGas.toString().padEnd(3, "");
    const percent = ((stat.totalGas / totalGas) * 100).toFixed(2);
    return `  ${name} | ${totalGasStr} gas | ${stat.totalHits} hits | ${stat.avgGas} avg gas | ${percent}%`;
};
