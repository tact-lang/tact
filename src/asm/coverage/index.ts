import { Cell } from "@ton/core";
import { compileCellWithMapping, decompileCell, Mapping } from "@/asm/runtime";
import { print, parse } from "@/asm/text";
import {
    createMappingInfo,
    createTraceInfoPerTransaction,
    loadFuncMapping,
} from "@/asm/trace";
import {
    buildFuncLineInfo,
    buildLineInfo,
    generateCoverageSummary,
} from "@/asm/coverage/data";
import { readFileSync } from "node:fs";

export function collectAsmCoverage(cell: Cell, logs: string) {
    const [cleanCell, mapping] = recompileCell(cell, false);
    const info = createMappingInfo(mapping);

    const traceInfos = createTraceInfoPerTransaction(logs, info, undefined);
    const assembly = print(decompileCell(cleanCell));
    const combinedTrace = { steps: traceInfos.flatMap((trace) => trace.steps) };
    const combinedLines = buildLineInfo(combinedTrace, assembly);
    const combinedSummary = generateCoverageSummary(combinedLines);
    return { lines: combinedLines, summary: combinedSummary };
}

export function collectFuncCoverage(
    cell: Cell,
    logs: string,
    funcSources: string,
    funcMappingPath: string,
) {
    const [_, mapping] = recompileCell(cell, true);
    const info = createMappingInfo(mapping);

    const funcMapping = loadFuncMapping(readFileSync(funcMappingPath, "utf8"));
    const traceInfos = createTraceInfoPerTransaction(logs, info, funcMapping);
    const func = readFileSync(funcSources, "utf8");
    const combinedLines = buildFuncLineInfo(traceInfos, func);
    const combinedSummary = generateCoverageSummary(combinedLines);
    return { lines: combinedLines, summary: combinedSummary };
}

export const recompileCell = (
    cell: Cell,
    forFunC: boolean,
): [Cell, Mapping] => {
    const instructionsWithoutPositions = decompileCell(cell);
    const assemblyForPositionsRaw = print(instructionsWithoutPositions);

    // filter out all DEBUGMARK lines from the assembly
    const assemblyForPositions = forFunC
        ? assemblyForPositionsRaw
        : assemblyForPositionsRaw
              .split("\n")
              .filter((it) => !it.includes("DEBUGMARK"))
              .join("\n");

    const parseResult = parse("out.tasm", assemblyForPositions);
    if (parseResult.$ === "ParseFailure") {
        throw new Error("Cannot parse resulting text Assembly");
    }

    return compileCellWithMapping(parseResult.instructions);
};

export * from "@/asm/coverage/html";
export * from "@/asm/coverage/text";
export * from "@/asm/coverage/integrations";
export * from "@/asm/coverage/data";
