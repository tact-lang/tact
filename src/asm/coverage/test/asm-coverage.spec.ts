import { compileCell } from "@/asm/runtime";
import { parse } from "@/asm/text/parse";
import { executeInstructions } from "@/asm/helpers/execute";
import { collectAsmCoverage } from "@/asm/coverage/index";
import { generateTextReport } from "@/asm/coverage/text";
import { generateHtml } from "@/asm/coverage/html";
import { mkdirSync, writeFileSync, existsSync } from "node:fs";
import { step } from "@/test/allure/allure";

describe("asm coverage", () => {
    const test =
        (name: string, code: string, id: number = 0) =>
        async () => {
            const res = parse("test.asm", code);
            if (res.$ === "ParseFailure") {
                throw new Error(res.error.msg);
            }

            const cell = compileCell(res.instructions);
            const [_, logs] = await executeInstructions(res.instructions, id);
            const { lines, summary } = collectAsmCoverage(cell, logs);

            const report = await step("Generating text report", () => {
                return generateTextReport(lines, summary);
            });
            await step("Report should match snapshot", () => {
                expect(report).toMatchSnapshot();
            });
            const outDirname = `${__dirname}/output`;
            if (!existsSync(outDirname)) {
                mkdirSync(outDirname);
            }

            const htmlReport = await step("Generating HTML report", () => {
                return generateHtml(lines);
            });
            await step("Write html report", () => {
                writeFileSync(`${__dirname}/output/${name}.html`, htmlReport);
            });
        };

    it(
        "simple if",
        test(
            "simple if",
            `
                PUSHINT 0
                PUSHINT 0
                PUSHCONT {
                    INC
                    INC
                    INC
                }
                IF
            `,
        ),
    );

    it(
        "if ret",
        test(
            "if ret",
            `
                DROP
                PUSHINT -1 // cond

                IFRET

                PUSHINT 1
                PUSHINT 2
                ADD
            `,
        ),
    );

    it(
        "simple if-else",
        test(
            "simple if-else",
            `
                PUSHINT 0
                PUSHINT -1
                PUSHCONT {
                    INC
                }
                PUSHCONT {
                    DEC
                }
                IFELSE
            `,
        ),
    );

    it(
        "while loop with break",
        test(
            "while loop with break",
            `
                PUSHINT 10 // a = 10

                PUSHCONT { DUP GTINT 0 } // a > 0
                PUSHCONT {
                    // if (a < 5) { break }
                    DUP
                    LESSINT 5
                    IFRETALT

                    // a -= 1;
                    DEC
                }
                WHILEBRK
            `,
        ),
    );

    it(
        "dictionary",
        test(
            "dictionary",
            `
                DICTPUSHCONST 19 [
                    0 => {
                        PUSHINT 10
                        INC
                    }
                    2 => {
                        PUSHINT 5
                        INC
                    }
                ]
                DICTIGETJMPZ
                THROWARG 11
            `,
        ),
    );

    it(
        "dictionary 2",
        test(
            "dictionary 2",
            `
                DICTPUSHCONST 19 [
                    0 => {
                        PUSHINT 10
                        INC
                    }
                    2 => {
                        PUSHINT 5
                        INC
                    }
                ]
                DICTIGETJMPZ
                THROWARG 11
            `,
            2,
        ),
    );

    it(
        "try without throw",
        test(
            "try without throw",
            `
                PUSHINT 10
                PUSHCONT {
                    INC
                }
                PUSHCONT {
                    DEC
                }
                TRY
            `,
        ),
    );

    it(
        "try with throw",
        test(
            "try with throw",
            `
                PUSHINT 10
                PUSHCONT {
                    THROW 10
                }
                PUSHCONT {
                    SUB
                }
                TRY
            `,
        ),
    );

    it(
        "nested try with rethrow",
        test(
            "nested try with rethrow",
            `
                PUSHCONT {
                    PUSHCONT {
                        THROW 10
                    }
                    PUSHCONT {
                        THROWANY
                    }
                    TRY
                }
                PUSHCONT {
                    PUSHINT 2
                }
                TRY
            `,
        ),
    );
});
