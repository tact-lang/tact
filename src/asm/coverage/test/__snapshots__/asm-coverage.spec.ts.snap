// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asm coverage dictionary 1`] = `
"Coverage Summary:
Lines: 4/8 (50.00%)
Total Gas: 283
Total Hits: 4

Instruction Stats:
  DICTIGETJMPZ    | 226 gas | 1 hits | 226 avg gas | 79.86%
  DICTPUSHCONST   | 34 gas | 1 hits | 34 avg gas | 12.01%
  PUSHINT         | 18 gas | 1 hits | 18 avg gas | 6.36%
  INC             | 5 gas | 1 hits | 5 avg gas | 1.77%

Annotated Code:
 1 ✓ | DICTPUSHCONST 19 [                       | gas:34    | hits:1
 2   |     0 => {                               |           |
 3 ✓ |         PUSHINT 10                       | gas:18    | hits:1
 4 ✓ |         INC                              | gas:5     | hits:1
 5   |     }                                    |           |
 6   |     2 => {                               |           |
 7 ✗ |         PUSHINT 5                        |           |
 8 ✗ |         INC                              |           |
 9   |     }                                    |           |
10 ✗ | ]                                        |           |
11 ✓ | DICTIGETJMPZ                             | gas:226   | hits:1
12 ✗ | THROWARG 11                              |           |
13   |                                          |           |"
`;

exports[`asm coverage dictionary 2 1`] = `
"Coverage Summary:
Lines: 4/8 (50.00%)
Total Gas: 283
Total Hits: 4

Instruction Stats:
  DICTIGETJMPZ    | 226 gas | 1 hits | 226 avg gas | 79.86%
  DICTPUSHCONST   | 34 gas | 1 hits | 34 avg gas | 12.01%
  PUSHINT         | 18 gas | 1 hits | 18 avg gas | 6.36%
  INC             | 5 gas | 1 hits | 5 avg gas | 1.77%

Annotated Code:
 1 ✓ | DICTPUSHCONST 19 [                       | gas:34    | hits:1
 2   |     0 => {                               |           |
 3 ✗ |         PUSHINT 10                       |           |
 4 ✗ |         INC                              |           |
 5   |     }                                    |           |
 6   |     2 => {                               |           |
 7 ✓ |         PUSHINT 5                        | gas:18    | hits:1
 8 ✓ |         INC                              | gas:5     | hits:1
 9   |     }                                    |           |
10 ✗ | ]                                        |           |
11 ✓ | DICTIGETJMPZ                             | gas:226   | hits:1
12 ✗ | THROWARG 11                              |           |
13   |                                          |           |"
`;

exports[`asm coverage if ret 1`] = `
"Coverage Summary:
Lines: 3/6 (50.00%)
Total Gas: 54
Total Hits: 3

Instruction Stats:
  DROP            | 18 gas | 1 hits | 18 avg gas | 33.33%
  PUSHINT         | 18 gas | 1 hits | 18 avg gas | 33.33%
  IFRET           | 18 gas | 1 hits | 18 avg gas | 33.33%

Annotated Code:
1 ✓ | DROP                                     | gas:18    | hits:1
2 ✓ | PUSHINT -1                               | gas:18    | hits:1
3 ✓ | IFRET                                    | gas:18    | hits:1
4 ✗ | PUSHINT 1                                |           |
5 ✗ | PUSHINT 2                                |           |
6 ✗ | ADD                                      |           |
7   |                                          |           |"
`;

exports[`asm coverage nested try with rethrow 1`] = `
"Coverage Summary:
Lines: 9/9 (100.00%)
Total Gas: 321
Total Hits: 9

Instruction Stats:
  PUSHCONT        | 104 gas | 4 hits | 26 avg gas | 32.40%
  THROW           | 84 gas | 1 hits | 84 avg gas | 26.17%
  THROWANY        | 76 gas | 1 hits | 76 avg gas | 23.68%
  TRY             | 52 gas | 2 hits | 26 avg gas | 16.20%
  PUSHINT         | 5 gas | 1 hits | 5 avg gas | 1.56%

Annotated Code:
 1 ✓ | PUSHCONT {                               | gas:26    | hits:1
 2 ✓ |     PUSHCONT {                           | gas:26    | hits:1
 3 ✓ |         THROW 10                         | gas:84    | hits:1
 4   |     }                                    |           |
 5 ✓ |     PUSHCONT {                           | gas:26    | hits:1
 6 ✓ |         THROWANY                         | gas:76    | hits:1
 7   |     }                                    |           |
 8 ✓ |     TRY                                  | gas:26    | hits:1
 9   | }                                        |           |
10 ✓ | PUSHCONT {                               | gas:26    | hits:1
11 ✓ |     PUSHINT 2                            | gas:5     | hits:1
12   | }                                        |           |
13 ✓ | TRY                                      | gas:26    | hits:1
14   |                                          |           |"
`;

exports[`asm coverage simple if 1`] = `
"Coverage Summary:
Lines: 4/7 (57.14%)
Total Gas: 67
Total Hits: 4

Instruction Stats:
  PUSHINT         | 36 gas | 2 hits | 18 avg gas | 53.73%
  PUSHCONT        | 26 gas | 1 hits | 26 avg gas | 38.81%
  IF              | 5 gas | 1 hits | 5 avg gas | 7.46%

Annotated Code:
1 ✓ | PUSHINT 0                                | gas:18    | hits:1
2 ✓ | PUSHINT 0                                | gas:18    | hits:1
3 ✓ | PUSHCONT {                               | gas:26    | hits:1
4 ✗ |     INC                                  |           |
5 ✗ |     INC                                  |           |
6 ✗ |     INC                                  |           |
7   | }                                        |           |
8 ✓ | IF                                       | gas:5     | hits:1
9   |                                          |           |"
`;

exports[`asm coverage simple if-else 1`] = `
"Coverage Summary:
Lines: 6/7 (85.71%)
Total Gas: 111
Total Hits: 6

Instruction Stats:
  PUSHCONT        | 52 gas | 2 hits | 26 avg gas | 46.85%
  PUSHINT         | 36 gas | 2 hits | 18 avg gas | 32.43%
  IFELSE          | 18 gas | 1 hits | 18 avg gas | 16.22%
  INC             | 5 gas | 1 hits | 5 avg gas | 4.50%

Annotated Code:
 1 ✓ | PUSHINT 0                                | gas:18    | hits:1
 2 ✓ | PUSHINT -1                               | gas:18    | hits:1
 3 ✓ | PUSHCONT {                               | gas:26    | hits:1
 4 ✓ |     INC                                  | gas:5     | hits:1
 5   | }                                        |           |
 6 ✓ | PUSHCONT {                               | gas:26    | hits:1
 7 ✗ |     DEC                                  |           |
 8   | }                                        |           |
 9 ✓ | IFELSE                                   | gas:18    | hits:1
10   |                                          |           |"
`;

exports[`asm coverage try with throw 1`] = `
"Coverage Summary:
Lines: 6/6 (100.00%)
Total Gas: 185
Total Hits: 6

Instruction Stats:
  THROW           | 84 gas | 1 hits | 84 avg gas | 45.41%
  PUSHCONT        | 52 gas | 2 hits | 26 avg gas | 28.11%
  TRY             | 26 gas | 1 hits | 26 avg gas | 14.05%
  PUSHINT         | 18 gas | 1 hits | 18 avg gas | 9.73%
  SUB             | 5 gas | 1 hits | 5 avg gas | 2.70%

Annotated Code:
1 ✓ | PUSHINT 10                               | gas:18    | hits:1
2 ✓ | PUSHCONT {                               | gas:26    | hits:1
3 ✓ |     THROW 10                             | gas:84    | hits:1
4   | }                                        |           |
5 ✓ | PUSHCONT {                               | gas:26    | hits:1
6 ✓ |     SUB                                  | gas:5     | hits:1
7   | }                                        |           |
8 ✓ | TRY                                      | gas:26    | hits:1
9   |                                          |           |"
`;

exports[`asm coverage try without throw 1`] = `
"Coverage Summary:
Lines: 5/6 (83.33%)
Total Gas: 101
Total Hits: 5

Instruction Stats:
  PUSHCONT        | 52 gas | 2 hits | 26 avg gas | 51.49%
  TRY             | 26 gas | 1 hits | 26 avg gas | 25.74%
  PUSHINT         | 18 gas | 1 hits | 18 avg gas | 17.82%
  INC             | 5 gas | 1 hits | 5 avg gas | 4.95%

Annotated Code:
1 ✓ | PUSHINT 10                               | gas:18    | hits:1
2 ✓ | PUSHCONT {                               | gas:26    | hits:1
3 ✓ |     INC                                  | gas:5     | hits:1
4   | }                                        |           |
5 ✓ | PUSHCONT {                               | gas:26    | hits:1
6 ✗ |     DEC                                  |           |
7   | }                                        |           |
8 ✓ | TRY                                      | gas:26    | hits:1
9   |                                          |           |"
`;

exports[`asm coverage while loop with break 1`] = `
"Coverage Summary:
Lines: 10/10 (100.00%)
Total Gas: 756
Total Hits: 45

Instruction Stats:
  DUP             | 252 gas | 14 hits | 18 avg gas | 33.33%
  LESSINT         | 182 gas | 7 hits | 26 avg gas | 24.07%
  IFRETALT        | 161 gas | 7 hits | 23 avg gas | 21.30%
  PUSHCONT        | 52 gas | 2 hits | 26 avg gas | 6.88%
  GTINT           | 35 gas | 7 hits | 5 avg gas | 4.63%
  DEC             | 30 gas | 6 hits | 5 avg gas | 3.97%
  WHILEBRK        | 26 gas | 1 hits | 26 avg gas | 3.44%
  PUSHINT         | 18 gas | 1 hits | 18 avg gas | 2.38%

Annotated Code:
 1 ✓ | PUSHINT 10                               | gas:18    | hits:1
 2 ✓ | PUSHCONT {                               | gas:26    | hits:1
 3 ✓ |     DUP                                  | gas:126   | hits:7
 4 ✓ |     GTINT 0                              | gas:35    | hits:7
 5   | }                                        |           |
 6 ✓ | PUSHCONT {                               | gas:26    | hits:1
 7 ✓ |     DUP                                  | gas:126   | hits:7
 8 ✓ |     LESSINT 5                            | gas:182   | hits:7
 9 ✓ |     IFRETALT                             | gas:161   | hits:7
10 ✓ |     DEC                                  | gas:30    | hits:6
11   | }                                        |           |
12 ✓ | WHILEBRK                                 | gas:26    | hits:1
13   |                                          |           |"
`;
